src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+        /// or if the current depth exceeds the recursive limit set by the max depth.
+        /// </exception>
+        public bool Read() => ReadSingleSegment();
+
+        private void StartObject()
+        {
+            _currentDepth++;
+            if (_currentDepth > _maxDepth)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ObjectDepthTooLarge);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+                _stackFreeContainer = (_stackFreeContainer << 1) | 1;
+            else
+                _stack.Push(JsonTokenType.StartObject);
+
+            _tokenType = JsonTokenType.StartObject;
+            _inObject = true;
+        }
+
+        private void EndObject()
+        {
+            if (!_inObject || _currentDepth <= 0)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ObjectEndWithinArray);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+            {
+                _stackFreeContainer >>= 1;
+                _inObject = (_stackFreeContainer & 1) != 0;
+            }
+            else
+            {
+                _inObject = _stack.Pop() != JsonTokenType.StartArray;
+            }
+
+            _currentDepth--;
+            _tokenType = JsonTokenType.EndObject;
+        }
+
+        private void StartArray()
+        {
+            _currentDepth++;
+            if (_currentDepth > _maxDepth)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ArrayDepthTooLarge);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+                _stackFreeContainer = _stackFreeContainer << 1;
+            else
+                _stack.Push(JsonTokenType.StartArray);
+
+            _tokenType = JsonTokenType.StartArray;
+            _inObject = false;
+        }
+
+        private void EndArray()
+        {
+            if (_inObject || _currentDepth <= 0)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ArrayEndWithinObject);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+            {
+                _stackFreeContainer >>= 1;
+                _inObject = (_stackFreeContainer & 1) != 0;
+            }
+            else
+            {
+                _inObject = _stack.Pop() != JsonTokenType.StartArray;
+            }
+
+            _currentDepth--;
+            _tokenType = JsonTokenType.EndArray;
+        }
+
+        private bool ReadSingleSegment()
+        {
+            bool retVal = false;
+            IsValueMultiSegment = false;
+
+            if (!HasMoreData())
+                goto Done;
```

Comment:
goto Done feels very Win32/C.  Why not just return and make it clear what is happening?

Reply:
This method is called in a tight-loop and is written this way as a performance optimization. Multiple return points regresses performance. I would be hesitant to change it in ways that might help readability but impacts perf significantly.